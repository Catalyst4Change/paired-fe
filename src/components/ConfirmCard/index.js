import React, { Component, Fragment } from 'react';
import { determineDisplayTime } from '../../helpers';
import PropTypes from 'prop-types';

export default class ConfirmCard extends Component {
  constructor() {
    super();
    this.state = {
      notes: ''
    };
  }

  handleChange = event => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };

  handleCancel = () => {
    const { history } = this.props;
    history.push('/book-pairing');
  };

  handleConfirm = () => {
    const { notes } = this.state;
    const {
      confirmPairing,
      deletePairingThunk,
      hasOpeningAlready,
      selectedPairing,
      userId,
      history
    } = this.props;
    confirmPairing(selectedPairing, userId, notes);
    if (hasOpeningAlready) {
      const { pairingInConflictId } = this.props;
      deletePairingThunk(pairingInConflictId);
    }
    history.push('/schedule');
  };

  render() {
    const { date, time, name, hasOpeningAlready } = this.props;

    return (
      <div className='ConfirmCard'>
        <h3>Booking {name}:</h3>
        <p>
          {date} at {determineDisplayTime(time)}
        </p>
        <label htmlFor='notes'>Tell {name} what you'd like to pair on: </label>
        <input
          name='notes'
          value={this.state.notes}
          onChange={this.handleChange}
        />
        {hasOpeningAlready && (
          <Fragment>
            <p>
              Please note you have an opening on your availability for this date
              and time
            </p>
            <p>Click confirm to continue and your opening will be removed</p>
            <p>Or click cancel to choose a different pairing</p>
          </Fragment>
        )}
        <button
          className='ConfirmCard--button--confirm'
          onClick={this.handleConfirm}>
          Confirm Booking
        </button>
        <button
          className='ConfirmCard--button--cancel'
          onClick={this.handleCancel}>
          Cancel
        </button>
      </div>
    );
  }
}

ConfirmCard.propTypes = {
  confirmPairing: PropTypes.func,
  date: PropTypes.string,
  deletePairingThunk: PropTypes.func,
  hasOpeningAlready: PropTypes.bool,
  history: PropTypes.object,
  name: PropTypes.string,
  selectedPairing: PropTypes.string,
  time: PropTypes.string,
  userId: PropTypes.string
};
