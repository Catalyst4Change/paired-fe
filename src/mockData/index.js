export const mockUser = {
  name: "Hillary",
  program: "FE",
  module: 4,
  id: "5caa7eebfdebb8348e53a48e"
}

export const mockSchedule = [
  {
    pairer: {
      name: "Hillary",
      module: 4,
      program: "FE",
      id: "5caa7eebfdebb8348e53a48e",
      pronouns: "she/her",
      slack: "@hillstew"
    },
    pairee: null,
    date: "Mon Apr 15 2019",
    time: "lunch",
    id: "5caa7eebfdebb8348e53a490",
    notes: null
  },
  {
    pairer: {
      name: "Hillary",
      module: 4,
      program: "FE",
      id: "5caa7eebfdebb8348e53a48e",
      pronouns: "she/her",
      slack: "@hillstew"
    },
    paireee: {
      name: "Jeo",
      module: 4,
      program: "FE",
      pronouns: "he/him",
      slack: "@jeo",
    },
    date: "Wed Apr 17 2019",
    time: "afternoon",
    id: "5caa7eebfdebb8348e53a49d",
    notes: null,
  }
];

export const mockAvailPairings = [
  {
    pairer: {
      name: "Tiffany",
      module: 4,
      program: "FE",
      id: "5caa7eebfdebb8348e53a48d",
      pronouns: "she/her",
      skills: [
        "react",
        "testing"
      ],
      slack: "@tiffany",
    },
    date: "Thu Apr 18 2019",
    time: "morning",
    id: "5caa7eebfdebb8348e53a4a4",
  },
  {
    pairer: {
      name: "Tiffany",
      module: 4,
      program: "FE",
      id: "5caa7eebfdebb8348e53a48d",
      pronouns: "she/her",
      skills: [
        "react",
        "testing"
      ],
      slack: "@tiffany",
    },
    date: "Thu Apr 18 2019",
    time: "lunch",
    id: "5caa7eebfdebb8348e53a4a5",
  }
]; 

export const mockPairingToDeleteFromSched = {
  pairer: {
    name: 'Hillary',
    module: 4,
    program: 'FE',
    id: '5caa7eebfdebb8348e53a48e',
    pronouns: 'she/her',
    slack: '@hillstew'
  },
  pairee: null,
  date: 'Mon Apr 15 2019',
  time: 'lunch',
  id: '5caa7eebfdebb8348e53a490',
  notes: null
};

export const mockScheduleAfterDelete = [
  {
    pairer: {
      name: "Hillary",
      module: 4,
      program: "FE",
      id: "5caa7eebfdebb8348e53a48e",
      pronouns: "she/her",
      slack: "@hillstew"
    },
    paireee: {
      name: "Jeo",
      module: 4,
      program: "FE",
      pronouns: "he/him",
      slack: "@jeo",
    },
    date: "Wed Apr 17 2019",
    time: "afternoon",
    id: "5caa7eebfdebb8348e53a49d",
    notes: null,
  }
]; 

export const mockPairingToAddToSched = {
  pairer: {
    name: 'Hillary',
    module: 4,
    program: 'FE',
    id: '5caa7eebfdebb8348e53a48e',
    pronouns: 'she/her',
    slack: '@hillstew'
  },
  pairee: null,
  date: 'Thu Apr 18 2019',
  time: 'morning',
  id: '5caa7eebfdebb8348e56f000',
  notes: null
};

export const mockScheduleAfterAdd = [
  {
    pairer: {
      name: "Hillary",
      module: 4,
      program: "FE",
      id: "5caa7eebfdebb8348e53a48e",
      pronouns: "she/her",
      slack: "@hillstew"
    },
    pairee: null,
    date: "Mon Apr 15 2019",
    time: "lunch",
    id: "5caa7eebfdebb8348e53a490",
    notes: null
  },
  {
    pairer: {
      name: "Hillary",
      module: 4,
      program: "FE",
      id: "5caa7eebfdebb8348e53a48e",
      pronouns: "she/her",
      slack: "@hillstew"
    },
    paireee: {
      name: "Jeo",
      module: 4,
      program: "FE",
      pronouns: "he/him",
      slack: "@jeo",
    },
    date: "Wed Apr 17 2019",
    time: "afternoon",
    id: "5caa7eebfdebb8348e53a49d",
    notes: null,
  },
  mockPairingToAddToSched
];

export const mockQueryFromgetAvailablePairings = {
         query: `{
    getAvailablePairings(filter: { program: "BE", module: 3, date: "Wed Apr 17 2019" }) {
      pairer {
        name
        module
        program
        id
        pronouns
        skills
        slack
      }
      date
      time
      id
    }
  }`
       };

export const mockQueryFromgetUser = {
         query: `{
  getUser(name: "Hillary") {
    name
    program
    module
    id
  }
}`
       };

export const mockQueryFromgetUserPairings = {
         query: `{
    getUserPairings(
      id: "5h5h5h5h5h5h5"
    ) {
      pairer {
        name
        module
        program
        id
        pronouns
        slack
      }
      pairee {
        name
        module
        program
        pronouns
        slack
      }
      date
      time
      id
      notes
    }
  }`
       };

export const mockQueryFromupdatePairing = {
         query: `mutation {
    updatePairing(pairing: {
        id: "6t6t6t"
        pairee: "7b7b7b"
        notes: "Mythical Creatures"
    }) {
      pairer {
        name
        module
        program
        id
        pronouns
        slack
      }
      pairee {
        name
        module
        program
        id
        pronouns
        slack
      }
      date
      time
      id
      notes
    }
  }`
       };

export const mockQueryFromdeletePairing = {
         query: `mutation {
    deletePairing(id: "45dtl") {
      date
    }
  }`
       };