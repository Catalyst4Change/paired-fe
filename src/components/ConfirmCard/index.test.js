import React from 'react';
import { shallow } from 'enzyme';
import ConfirmCard from './index';
import * as data from '../../mockData';

describe('ConfirmCard', () => {
  let wrapper;
  const mockPairing = data.mockPairingForConfirmCard;
  const mockHistoryPush = jest.fn();
  const mockConfirmPairing = jest.fn();
  const mockDeletePairingThunk = jest.fn();
  const mockProps = {
    date: mockPairing.date,
    time: mockPairing.time,
    name: mockPairing.pairer.name,
    selectedPairing: mockPairing.id,
    hasOpeningAlready: false,
    history: { push: mockHistoryPush },
    confirmPairing: mockConfirmPairing,
    deletePairingThunk: mockDeletePairingThunk,
    userId: data.mockUser.id
  };
  const mockPairingInConflictId = '49msfj';
  const mockPropsWithConflict = {
    ...mockProps,
    hasOpeningAlready: true,
    pairingInConflictId: mockPairingInConflictId
  };

  beforeEach(() => {
    wrapper = shallow(<ConfirmCard {...mockProps} />);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the snapshot if the user has an opening already at the same time', () => {
    const wrapper = shallow(<ConfirmCard {...mockPropsWithConflict} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should have a default state', () => {
    const expected = {
      notes: ''
    };
    expect(wrapper.state()).toEqual(expected);
  });

  describe('handleChange', () => {
    it('should call handleChange when a user types in the input', () => {
      const instance = wrapper.instance();
      jest.spyOn(instance, 'handleChange');
      wrapper.instance().forceUpdate();
      wrapper.find('input').simulate('change', {
        target: { name: 'notes', value: 'Help' }
      });
      expect(instance.handleChange).toHaveBeenCalled();
    });

    it('should update notes in state when user enters notes in input', () => {
      const mockUserNotes = 'Help with grid';
      wrapper.find('input').simulate('change', {
        target: { name: 'notes', value: mockUserNotes }
      });
      expect(wrapper.state('notes')).toEqual(mockUserNotes);
    });
  });

  describe('handleCancel', () => {
    it('should call handleCancel when a user clicks cancel btn', () => {
      const instance = wrapper.instance();
      jest.spyOn(instance, 'handleCancel');
      wrapper.instance().forceUpdate();
      wrapper.find('button.ConfirmCard--button--cancel').simulate('click');
      expect(instance.handleCancel).toHaveBeenCalled();
    });

    it('should redirect user to /book-pairing when cancel is clicked', () => {
      const expectedPath = '/book-pairing';
      wrapper.find('button.ConfirmCard--button--cancel').simulate('click');
      expect(mockHistoryPush).toHaveBeenCalledWith(expectedPath);
    });
  });

  describe('handleConfirm', () => {
    it('should call handleConfirm when a user clicks confirm btn', () => {
      const instance = wrapper.instance();
      jest.spyOn(instance, 'handleConfirm');
      wrapper.instance().forceUpdate();
      wrapper.find('button.ConfirmCard--button--confirm').simulate('click');
      expect(instance.handleConfirm).toHaveBeenCalled();
    });

    it('should call confirmPairing with the correct params', () => {
      const mockNotes = 'JSFun problem set';
      wrapper.setState({ notes: mockNotes });
      wrapper.instance().handleConfirm();
      expect(mockConfirmPairing).toHaveBeenCalledWith(
        mockPairing.id,
        data.mockUser.id,
        mockNotes
      );
    });

    it('should redirect user to / ', () => {
      const expectedPath = '/';
      wrapper.instance().handleConfirm();
      expect(mockHistoryPush).toHaveBeenCalledWith(expectedPath);
    });

    it('should call deletePairingThunk is they already have an open paring', () => {
      const wrapper = shallow(<ConfirmCard {...mockPropsWithConflict} />);
      wrapper.instance().handleConfirm();
      expect(mockDeletePairingThunk).toHaveBeenCalledWith(mockPairingInConflictId)
    });
  });
});
