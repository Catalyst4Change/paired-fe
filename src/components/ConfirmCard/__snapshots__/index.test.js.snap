// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmCard should match the snapshot 1`] = `
<div
  className="ConfirmCard"
>
  <h3>
    Booking 
    Tiffany
    :
  </h3>
  <p>
    Fri Apr 07 2019
     at 
    12:00 - 12:50 p.m.
  </p>
  <label
    htmlFor="notes"
  >
    Tell 
    Tiffany
     what you'd like to pair on:
     
  </label>
  <input
    name="notes"
    onChange={[Function]}
    value=""
  />
  <button
    className="ConfirmCard--button--confirm"
    onClick={[Function]}
  >
    Confirm Booking
  </button>
  <button
    className="ConfirmCard--button--cancel"
    onClick={[Function]}
  >
    Cancel
  </button>
</div>
`;

exports[`ConfirmCard should match the snapshot if the user has an opening already at the same time 1`] = `
<div
  className="ConfirmCard"
>
  <h3>
    Booking 
    Tiffany
    :
  </h3>
  <p>
    Fri Apr 07 2019
     at 
    12:00 - 12:50 p.m.
  </p>
  <label
    htmlFor="notes"
  >
    Tell 
    Tiffany
     what you'd like to pair on:
     
  </label>
  <input
    name="notes"
    onChange={[Function]}
    value=""
  />
  <p>
    Please note you have an opening on your availability for this date and time
  </p>
  <p>
    Click confirm to continue and your opening will be removed
  </p>
  <p>
    Or click cancel to choose a different pairing
  </p>
  <button
    className="ConfirmCard--button--confirm"
    onClick={[Function]}
  >
    Confirm Booking
  </button>
  <button
    className="ConfirmCard--button--cancel"
    onClick={[Function]}
  >
    Cancel
  </button>
</div>
`;
